launcher/updatedialog_moc.cpp

```diff
@@ -0,0 +1,143 @@
+	
+	if(settings["launcher"]["updateOnStartup"].Bool() == true)
+		ui->checkOnStartup->setCheckState(Qt::CheckState::Checked);
+	
+	currentVersion = GameConstants::VCMI_VERSION;
+	
+	setWindowTitle(QString::fromStdString(currentVersion));
+	
+#ifdef VCMI_WINDOWS
+	platformParameter = "windows";
+#elif defined(VCMI_MAC)
+	platformParameter = "macos";
+#elif defined(VCMI_IOS)
+	platformParameter = "ios";
+#elif defined(VCMI_ANDROID)
+	platformParameter = "android";
+#elif defined(VCMI_UNIX)
+	platformParameter = "linux";
+#endif
+	
+	QString url = QString::fromStdString(settings["launcher"]["updateConfigUrl"].String());
+	
+	QScopedPointer<QNetworkAccessManager> manager(new QNetworkAccessManager);
+	
+	QNetworkReply *response = manager->get(QNetworkRequest(QUrl(url)));
+	
+	QObject::connect(response, &QNetworkReply::finished, [&, response]{
+		response->deleteLater();
+		response->manager()->deleteLater();
+		if(response->error() != QNetworkReply::NoError)
+		{
+			ui->versionLabel->setStyleSheet("QLabel { background-color : red; color : black; }");
+			ui->versionLabel->setText("Network error");
+			ui->plainTextEdit->setPlainText(response->errorString());
+			return;
+		}
+		
+		auto byteArray = response->readAll();
+		JsonNode node(byteArray.constData(), byteArray.size());
+		loadFromJson(node);
+		
+	}) && manager.take();
+}
+
+UpdateDialog::~UpdateDialog()
+{
+	delete ui;
+}
+
+void UpdateDialog::showUpdateDialog(bool isManually)
+{
+	UpdateDialog * dialog = new UpdateDialog(nullptr, isManually);
+	
+	dialog->setAttribute(Qt::WA_DeleteOnClose);
+}
+
+void UpdateDialog::on_checkOnStartup_stateChanged(int state)
+{
+	Settings node = settings.write["launcher"]["updateOnStartup"];
+	node->Bool() = (state == 2 ? true : false);
+}
+
+void UpdateDialog::loadFromJson(const JsonNode & node)
+{
+	if(node.getType() != JsonNode::JsonType::DATA_STRUCT ||
+	   node["updateType"].getType() != JsonNode::JsonType::DATA_STRING ||
+	   node["version"].getType() != JsonNode::JsonType::DATA_STRING ||
+	   node["changeLog"].getType() != JsonNode::JsonType::DATA_STRING ||
+	   node.getType() != JsonNode::JsonType::DATA_STRUCT) //we need at least one link - other are optional
+	{
+		ui->plainTextEdit->setPlainText("Cannot read JSON from url or incorrect JSON data");
+		return;
+	}
+	
+	//check whether update is needed
+	std::string newVersion = node["version"].String();
+	if(currentVersion == newVersion)
+	{
+		if(!calledManually)
+			close();
+		
+		return;
+	}
+	
+	if(!calledManually)
+	{
+		setWindowModality(Qt::ApplicationModal);
+		show();
+	}
+	
+	if(node["updateType"].String() == "minor")
+		ui->versionLabel->setStyleSheet("QLabel { background-color : gray; color : black; }");
+	if(node["updateType"].String() == "major")
+		ui->versionLabel->setStyleSheet("QLabel { background-color : orange; color : black; }");
+	if(node["updateType"].String() == "critical")
+		ui->versionLabel->setStyleSheet("QLabel { background-color : red; color : black; }");
```

Comment:
```suggestion
	QLatin1String bgColor;
	const auto updateType = node["updateType"].String();
	if(updateType == "minor")
		bgColor = {"gray"};
	else if(updateType == "major")
		bgColor = {"orange"};
	else if(updateType == "critical")
		bgColor = {"red"};
	ui->versionLabel->setStyleSheet(QLatin1String("QLabel { background-color : %1; color : black; }").arg(bgColor));
```

can there be other `updateType`s?

Reply:
good point, actually it will be green by default

Reply:
btw code is not compilable

Reply:
> btw code is not compilable

and the error is?..

sorry, I typed it on GitHub directly.